x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/runtimefields/RuntimeFieldsFeatureSetUsage.java

```diff
@@ -8,283 +8,43 @@
-                    if (runtimeFieldMappingObject instanceof Map == false) {
-                        continue;
-                    }
-                    Map<?, ?> runtimeFieldMapping = (Map<?, ?>) runtimeFieldMappingObject;
-                    Object typeObject = runtimeFieldMapping.get("type");
-                    if (typeObject == null) {
-                        continue;
-                    }
-                    String type = typeObject.toString();
-                    RuntimeFieldStats stats = fieldTypes.computeIfAbsent(type, RuntimeFieldStats::new);
-                    stats.count++;
-                    if (indexFieldTypes.add(type)) {
-                        stats.indexCount++;
-                    }
-                    if (concreteFieldNames.contains(runtimeFieldName)) {
-                        stats.shadowedCount++;
-                    }
-                    Object scriptObject = runtimeFieldMapping.get("script");
-                    if (scriptObject == null) {
-                        stats.scriptLessCount++;
-                    } else if (scriptObject instanceof Map) {
-                        Map<?, ?> script = (Map<?, ?>) scriptObject;
-                        Object sourceObject = script.get("source");
-                        if (sourceObject != null) {
-                            String scriptSource = sourceObject.toString();
-                            int chars = scriptSource.length();
-                            long lines = scriptSource.split("\\n").length;
-                            int docUsages = countOccurrences(scriptSource, "doc[\\[\\.]");
-                            int sourceUsages = countOccurrences(scriptSource, "params\\._source");
-                            stats.update(chars, lines, sourceUsages, docUsages);
-                        }
-                        Object langObject = script.get("lang");
-                        if (langObject != null) {
-                            stats.scriptLangs.add(langObject.toString());
-                        }
-                    }
-                }
-            }
-        }
-        List<RuntimeFieldStats> runtimeFieldStats = new ArrayList<>(fieldTypes.values());
-        runtimeFieldStats.sort(Comparator.comparing(RuntimeFieldStats::type));
-        return new RuntimeFieldsFeatureSetUsage(Collections.unmodifiableList(runtimeFieldStats));
-    }
-
-    private final List<RuntimeFieldStats> stats;
-
-    RuntimeFieldsFeatureSetUsage(List<RuntimeFieldStats> stats) {
-        super(XPackField.RUNTIME_FIELDS, true, true);
-        this.stats = stats;
-    }
+    static final Version MINIMAL_SUPPORTED_VERSION = Version.V_7_11_0;
 
     public RuntimeFieldsFeatureSetUsage(StreamInput in) throws IOException {
-        super(in);
-        this.stats = in.readList(RuntimeFieldStats::new);
+        super(in.getVersion().before(Version.V_7_13_0) ? in.readString() : "runtime_fields",
```

Comment:
I'm wondering why we need to specialize the logic to the version. Could we adopt a simpler approach of always reading/ writing the parameters?

Reply:
good one, I guess I initially went for conditionally serializing just to make sure, but based on my understanding of how xpack usage works, this should not be needed. The important bit is that we never go and serialize this class to 8.0 which won't have the named writeable registered, but knowing that only 7.11 and 7.12 nodes may cause its serialization which are not 7.last, we should be ok because 7.11 and 7.12 will never be in the same cluster as 8.0 nodes. That said keeping the named writeable around for the whole 7.x series should do the trick. We could make it a no-op when this class gets serialized to/from 7.13+ but that should make no difference. Don't take my word on this though, were you thinking along the same lines? :)

Reply:
Yes this is exactly what I was thinking. For `flattened` fields we just kept the class around for the 7.x series, without any changes to its serialization logic.